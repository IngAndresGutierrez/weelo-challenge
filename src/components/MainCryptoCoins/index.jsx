/* import internal modules */
import { useEffect, useState } from 'react'

/* import internal modules */
import Loading from '../common/Loading'
import TableCryptoCoins from '../TableCryptoCoins'
import { Paper, Tab, Tabs } from '@material-ui/core'
import { getCryptoCoins } from '../../apis/cryptoCoins'
import DataGridCryptoCoins from '../DataGridCryptoCoins'
import { cryptoCoinsList } from '../../utils/cryptoCoinsList'

/** @description This is a functional component for layout cryptocurreny coins
 *  @version 1.0.0
 *  @since 11/14/2021
 *  @author Jaime Andrés Gómez Gutiérrez <g.gutierrez.j.andres@gmail.com>
 *
 *  @function
 *  @name MainCryptoCoins
 *  @returns {Component} Returns the layout of cryptocurreny coins component
 **/

const MainCryptoCoins = () => {
  const [value, setValue] = useState(0)
  const [loading, setLoading] = useState(true)
  const [coinsList, setCoinsList] = useState([])

  const handleChange = (event, newValue) => {
    setValue(newValue)
  }

  useEffect(() => {
    getCryptoCoinsFunction()
  }, [])

  const getCryptoCoinsFunction = () => {
    getCryptoCoins()
      .then((response) => {
        /* request success */
        if (response.status >= 200 && response.status <= 299) {
          setCoinsList(response.data.data)
          setLoading(false)
        }

        /* if there are any errors in the api server (only for this challenge) */
        if (response.status > 299) {
          setCoinsList(cryptoCoinsList)
          setLoading(false)
        }
      })
      .catch((error) => {
        console.error(error)

        /* if there are any errors in the api server (only for this challenge) */
        if (error) {
          setCoinsList(cryptoCoinsList)
          setLoading(false)
        }
      })
  }

  return (
    <>
      <Paper square>
        <Tabs
          value={value}
          textColor="primary"
          onChange={handleChange}
          indicatorColor="primary"
          aria-label="disabled tabs example"
        >
          <Tab label="Datagrid Crypto Coins" />
          <Tab label="Table Crypto Coins" />
        </Tabs>
      </Paper>
      {!loading ? (
        <>
          {value === 0 ? (
            <DataGridCryptoCoins coinsList={coinsList} />
          ) : (
            <TableCryptoCoins coinsList={coinsList} />
          )}
        </>
      ) : (
        <Loading />
      )}
    </>
  )
}

export default MainCryptoCoins
