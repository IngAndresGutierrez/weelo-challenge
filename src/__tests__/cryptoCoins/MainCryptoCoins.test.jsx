/* eslint-disable jest/valid-expect */
/* import external modules */
import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'

/* import internal modules */
import '../../../setupTests'
import MainCryptoCoins from '../../components/MainCryptoCoins'

describe('Testing suite MainCryptoCoins component', () => {
  const MainCryptoCoinsShallow = shallow(<MainCryptoCoins />)
  const MainCryptoCoinsMount = mount(<MainCryptoCoins />)

  test('Render MainCryptoCoins', () => {
    expect(MainCryptoCoinsShallow).toHaveLength(1)
  })

  test('Renders correctly - Snapshot', () => {
    const treeMainCryptoCoinsMount = renderer
      .create(MainCryptoCoinsMount)
      .toJSON()
    expect(treeMainCryptoCoinsMount).toMatchSnapshot()
  })

  test('Render information', () => {
    expect(MainCryptoCoinsMount.find('span').first().text()).toEqual(
      'Datagrid Crypto Coins'
    )
  })

  test('Render components', () => {
    expect(MainCryptoCoinsMount.contains('Paper'))
    expect(MainCryptoCoinsMount.contains('Tabs'))
    expect(MainCryptoCoinsMount.contains('DataGridCryptoCoins'))
    expect(MainCryptoCoinsMount.contains('TableCryptoCoins'))
  })
})
