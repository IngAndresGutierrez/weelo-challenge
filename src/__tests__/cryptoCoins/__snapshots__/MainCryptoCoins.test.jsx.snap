// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing suite MainCryptoCoins component Renders correctly - Snapshot 1`] = `
Array [
  <div
    className="MuiPaper-root MuiPaper-elevation1"
  >
    <div
      className="MuiTabs-root"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": null,
            "overflow": "hidden",
          }
        }
      >
        <div
          aria-label="disabled tabs example"
          className="MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Datagrid Crypto Coins
            </span>
            <span
              className="PrivateTabIndicator-root-1 PrivateTabIndicator-colorPrimary-2 MuiTabs-indicator"
              style={Object {}}
            />
          </button>
          <button
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Table Crypto Coins
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="makeStyles-root-5"
  >
    <div
      className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>,
]
`;
